Adding new documentation article on writing tests (#11604)
conda/conda: docs/source/dev-guide/writing-tests.md

# Writing Tests

The following guide gives an overview of writing tests for the `conda` codebase.
This article is mostly aimed at new contributors, especially those wishing

> to submit bug fixes. To give those new to writing tests in the conda codebase

Reviewer: ```suggestion
to submit bug fixes. To give those new to writing tests in the `conda` codebase
```

Author: @kathatherine package manager. Not the command. So, we probably just want to leave it as is, right?

    out, err, exit_code = run(f"conda remove --all -n {TEST_ENV_NAME_1}")

    assert err == ""
    assert exit_code == 0
```

Let's break down exactly what is going on in the code snippet above:

First, we import a function called `run_inprocess_conda_command` (aliased to `run` here) that allows
us to run a command using the current running process. This ends up being much more efficient and quicker than
running this test as a subprocess.

In the test itself, we first use our `run` function to create a new environment. This function
returns the standard out, standard error and the exit code of the command. This allows us to
perform our inspections in order to determine whether the command successfully ran.

The second part of the test again uses the `run` command to call `conda env list`. This time,
we pass the `--json` flag, which allows capturing JSON that we can better parse and more easily
inspect. This second part of the integration test not only ensures that our previous `conda create`
command ran successfully, but also tests to make sure that `conda env list` runs correctly.

Finally, we destroy the environment we just created and ensure the standard error and the exit

> code are what we expect them to be. It is important to remember removing anything you create
> as this will be present when other tests are run.

Reviewer 2: ```suggestion
code are what we expect them to be. It is important to remember to remove anything you create,
as it will be present when other tests are run.
```

### Tests with fixtures

Sometimes in integration tests, you may want to re-use the same type of environment more than once.
Copying and pasting this setup and teardown code into each individual test can make these tests more

> difficult read and harder to maintain.

Reviewer 2: ```suggestion
difficult to read and harder to maintain.
```

To overcome this, `conda` tests make extensive use of `pytest` fixtures. Below is an example of the
previously shown test except that we now make the focus of the test the `conda env list` command and move
the creation and removal of the environment into a fixture:

```python
# Writing a test for `conda env list`

import os.path
import json

import pytest

from conda.testing.helpers import run_inprocess_conda_command as run

TEST_ENV_NAME_1 = "test-env-1"

@pytest.fixture()
def env_one():
    out, err, exit_code = run(f"conda create -n {TEST_ENV_NAME_1} -y")

    assert exit_code == 0

    yield

    out, err, exit_code = run(f"conda remove --all -n {TEST_ENV_NAME_1}")

    assert exit_code == 0

def test_env_list_finds_existing_environment(env_one):

>     # Because we're using fixtures we can immediately run the `conda env list` command

Reviewer 2: ```suggestion
    # Because we're using fixtures, we can immediately run the `conda env list` command
```

    # and our test assertions
    out, err, exit_code = run("conda env list --json")
    json_out = json.loads(out)
    env_names = {os.path.basename(path) for path in json_out.get("envs", tuple())}

    assert TEST_ENV_NAME_1 in env_names
    assert err == ""
    assert exit_code == 0
```

> In the fixture named `env_one`, we first create a new environment exactly the same as we

Reviewer 2: ```suggestion
In the fixture named `env_one`, we first create a new environment in exactly the same way as we
```

did in our previous test. We make an assertion to ensure that it ran correctly, and
yield to mark the end of the setup. In the teardown section after the yield statement,

> we run the `conda remove` command and also make an assertion to determine in ran correctly.

Reviewer 2: ```suggestion
we run the `conda remove` command and also make an assertion to determine it ran correctly.
```

> This fixture will be run using the default scope in `pytest` which is `function`. This means
> that the setup and teardown will be run before and after each test. If you needed to share
> and environment or other pieces of data between tests, just remember to set the fixture

Reviewer 2: ```suggestion
This fixture will be run using the default scope in `pytest`, which is `function`. This means
that the setup and teardown will be run before and after each test. If you need to share
an environment or other pieces of data between tests, just remember to set the fixture
```
